src/ledger-iframe-bridge.ts

```diff
@@ -0,0 +1,291 @@
+import {
+  GetPublicKeyParams,
+  GetPublicKeyResponse,
+  LedgerBridge,
+  LedgerSignMessageParams,
+  LedgerSignMessageResponse,
+  LedgerSignTransactionParams,
+  LedgerSignTransactionResponse,
+  LedgerSignTypedDataParams,
+  LedgerSignTypedDataResponse,
+} from './ledger-bridge';
+
+const LEDGER_IFRAME_ID = 'LEDGER-IFRAME';
+
+export enum IFrameMessageAction {
+  LedgerConnectionChange = 'ledger-connection-change',
+  LedgerUnlock = 'ledger-unlock',
+  LedgerMakeApp = 'ledger-make-app',
+  LedgerUpdateTransport = 'ledger-update-transport',
+  LedgerSignTransaction = 'ledger-sign-transaction',
+  LedgerSignPersonalMessage = 'ledger-sign-personal-message',
+  LedgerSignTypedData = 'ledger-sign-typed-data',
+}
+
+type IFrameMessageResponse<TAction extends IFrameMessageAction> = {
+  action: TAction;
+  messageId: number;
+} & (
+  | {
+      action: IFrameMessageAction.LedgerConnectionChange;
+      payload: { connected: boolean };
+    }
+  | ({
+      action: IFrameMessageAction.LedgerMakeApp;
+    } & ({ success: true } | { success: false; error?: unknown }))
+  | {
+      action: IFrameMessageAction.LedgerUpdateTransport;
+      success: boolean;
+    }
+  | ({
+      action: IFrameMessageAction.LedgerUnlock;
+    } & (
+      | { success: true; payload: GetPublicKeyResponse }
+      | { success: false; payload: { error: Error } }
+    ))
+  | ({
+      action: IFrameMessageAction.LedgerSignTransaction;
+    } & (
+      | { success: true; payload: LedgerSignTransactionResponse }
+      | { success: false; payload: { error: Error } }
+    ))
+  | ({
+      action:
+        | IFrameMessageAction.LedgerSignPersonalMessage
+        | IFrameMessageAction.LedgerSignTypedData;
+    } & (
+      | {
+          success: true;
+          payload: LedgerSignMessageResponse | LedgerSignTypedDataResponse;
+        }
+      | { success: false; payload: { error: Error } }
+    ))
+);
+
+type IFrameMessage<TAction extends IFrameMessageAction> = {
+  action: TAction;
+  params?: Readonly<Record<string, unknown>>;
+};
+
+type IFramePostMessage<TAction extends IFrameMessageAction> =
+  IFrameMessage<TAction> & {
+    messageId: number;
+    target: typeof LEDGER_IFRAME_ID;
+  };
+
+export class LedgerIframeBridge implements LedgerBridge {
+  iframe?: HTMLIFrameElement;
+
+  iframeLoaded = false;
+
+  bridgeUrl = '';
+
+  isDeviceConnected = false;
+
+  currentMessageId = 0;
+
+  messageCallbacks: Record<
+    number,
+    (response: IFrameMessageResponse<IFrameMessageAction>) => void
+  > = {};
+
+  delayedPromise?: {
+    resolve: (value: boolean) => void;
+    reject: (error: unknown) => void;
+    transportType: string;
+  };
+
+  async init(bridgeUrl: string) {
+    this.bridgeUrl = bridgeUrl;
+
+    this.#setupIframe();
```

Comment:
Now that this is not in the constructor, but on an async method, it should be perfectly possible to await this method by returning a promise that resolves on `iframe.onLoad`. That way we can remove all the `iframeLoaded` and `delayedPromise` references and code.

I chose not to do it in this PR so that it's easier to review, but it's a very simply change afterwards.